src/common/Throttle.cc

```diff
@@ -668,26 +668,73 @@ uint64_t TokenBucketThrottle::Bucket::get(uint64_t c) {
   return got;
 }
 
-uint64_t TokenBucketThrottle::Bucket::put(uint64_t c) {
-  if (0 == max) {
+uint64_t TokenBucketThrottle::Bucket::put() {
+  if (0 == committed_information_rate) {
     return 0;
   }
-  if (c) {
-    // put c tokens into bucket
-    uint64_t current = remain;
-    if ((current + c) <= (uint64_t)max) {
-      remain += c;
-    } else {
-      remain = (uint64_t)max;
+
+  uint64_t c = tokens_per_tick;
+  uint64_t max = committed_burst_size;
+
+  // if Burst is disabled, the maximum is tokens_per_tick.
+  if (0 == max) {
+    max = tokens_per_tick;
+  }
+
+  // for the number(CIR) cannot be divisible
+  if (0 == divide_left) {
+    if (0 == ticks_per_second) {
+      ticks_per_second = 1000 / tick;
+      divide_left = committed_information_rate % ticks_per_second;
     }
+    ticks_per_second--;
+  } else {
+    divide_left--;
+    ticks_per_second--;
+    c++;
+    max++;
+  }
+
+  // put tokens into bucket
+  if (remain + c > max) {
+    remain = max;
+  } else {
+    remain += c;
   }
   return remain;
 }
 
+bool TokenBucketThrottle::Bucket::set_average(uint64_t c) {
+  if (0 == c) {
+    return false;
+  }
+  // calc the tick(ms), don't less than the minimum.
+  tick = 1000 / c;
+  if (tick < tick_min) {
+    tick = tick_min;
+  }
+
+  // tokens to put into the bucket every tick.
+  tokens_per_tick = c * tick / 1000;
+
+  // this is for the number(CIR) can not be divisible.
+  ticks_per_second = 1000 / tick;
+  divide_left = c % ticks_per_second;
+
+  // zero means Bursting IO is disabled.
+  if (0 == committed_burst_size) {
+    committed_burst_size = tokens_per_tick;
+    remain = tokens_per_tick;
+  } else {
+    remain = committed_burst_size;
+  }
+
+  committed_information_rate = c;
+  return true;
+}
+
 void TokenBucketThrottle::Bucket::set_max(uint64_t m) {
-  if (remain > m || max == 0)
-    remain = m;
-  max = m;
+  committed_burst_size = m;
```

Comment:
We need to update remain, if remain is too large here. For example, if the max is 100, average is 10 and remain is 100. that means the burst size is 100. But if we wan to set the burst size to 10. average to 1. If you don't set the remain to 10, user can still get an bursting 100 tokens.